### ISSUE TITLE
missing information on how to start

### ISSUE THREAD

**Issue by philippkeller** [koute/pinky#1]
I saw your project on the changelog weekly newsletter and wanted to try it out. But now I have no idea how to start. I downloaded retroarch but there's no cli coming with it. Can you flesh this out in your documentation?

**Comment by koute** [comment]
AFAIK RetroArch launches the GUI by default, but if you run `retroarch --help` it should print out what parameters it accepts; does that not work?

Yeah, I'll definitely have to improve the documentation here a little bit.

**Comment by philippkeller** [comment]
I expected that RetroArch puts a binary (or a symbolic link) into the path. But it seems it doesn't.

I'm on OSX and istalled with `brew cask install retroarch`. If I do this (from the pinky-libretro subdir):

`/Applications/RetroArch.app/Contents/MacOS/RetroArch -L libpinky_libretro.so ../nes-testsuite/roms/apu_test/1-len_ctr.nes`

then nothing happens. Where should `libpinky_libretro.so` be? It was not built after `cargo build`. I thought that `cargo install` would install the built files into the right directories but it exited with `error: no packages found with binaries or examples`.

Can you point me into the right direction?

**Comment by koute** [comment]
If you run `cargo build` in the `pinky-libretro` directory then you should get a shared library in `pinky-libretro/target/debug`; just please note that on systems other than Linux the shared object might have a different extension. (On OS X it'll probably be '.dylib', though I don't know for sure as I never compiled anything on OS X.)

Yes, `cargo install` won't do anything as the emulator isn't a standalone binary.

Let me know if you manage to run it successfully!

**Comment by philippkeller** [comment]
Yes, I got it working thanks to your hints. But it's a bit strange:

I first needed to run `/Applications/RetroArch.app/Contents/MacOS/RetroArch -L target/debug/libpinky_libretro.dylib ~/Desktop/mario.nes` and then the command "hangs" (seems to do nothing). Also when I start RetroArch "in parallel" without any flags the game loads and seems playable

**Comment by koute** [comment]
That is strange; what version of RetroArch do you have installed?

Also, do you have any other cores installed for RetroArch? If you launch RetroArch without any flags then it might use another NES emulator core that you've installed along with RetroArch.

Unfortunately I don't have access to a system with OS X to test whenever everything works on it or not. (I know it should work fine on Linux with RetroArch v1.3.4 since that's what I'm using.)

Could you run RetroArch with `-v --log-file=logs.txt` and paste me/send me the logs?

Thanks!

**Comment by philippkeller** [comment]
when running with `--log-file=logs.txt` it does a segmentation fault. It seems that OSX is not in focus for RetroArch but is kindof supported "on the side".

Any way, attached is the output of the process I started from command line (sending stderr to the logfile via `2>...`). Instead of just opening the window it pauses and only continues when starting RetroArch via the Finder (i.e. going over the mac ui instead of the command line).
[retroarch.txt](https://github.com/koute/pinky/files/560720/retroarch.txt)

**Comment by philippkeller** [comment]
to your other question: no, I didn't install any other core and RetroArch didn't come with any core. If I start RetroArch (without starting it in command line before with the -L argument) it shows me a screen where I need to load a core

**Comment by philippkeller** [comment]
can you recommend me any other libretro frontend? A short google search didn't lead me to any

**Comment by koute** [comment]
Wow, really? A segmentation fault? That's just sad.

I don't see anything obvious in the logs you've pasted; I wonder if the black screen issue is with RetroArch or with Pinky?

Some cursory Googling revealed [Phoenix](https://github.com/team-phoenix/Phoenix) as another libretro frontend; [it seems to have OS X binaries available](https://github.com/team-phoenix/Phoenix/releases). I've grabbed the Linux binaries and ran:

```
./Phoenix --libretro -c libpinky_libretro.so -g game.nes
```

and it seems to work pretty well! It's certainly a very pretty frontend, though it feels a little more sluggish than RetroArch. Maybe it'll work for you too?

**Comment by philippkeller** [comment]
I just tried it but with no lock either :-( things don't seem so well in the OSX world. I've attached the output of Phoenix.

> I wonder if the black screen issue is with RetroArch or with Pinky?
actually it's not a "black screen" it's just that the start from the command line does not work (just hangs) but needs a parallel start triggered from UI
[phoenix.txt](https://github.com/koute/pinky/files/561695/phoenix.txt)

**Comment by koute** [comment]
Hmmm... it's suspicious that two different libretro frontends don't work properly. Perhaps the issue is on my end, only it somehow ends up working on Linux but not on OS X?

Could you try running both frontends (RetroArch and Phoenix) from the command line but with a core from a different emulator? If that ends up working properly then it means that my libretro bindings are somehow broken.

**Comment by philippkeller** [comment]
late reply: I've tried with another core:

`/Applications/RetroArch.app/Contents/MacOS/RetroArch -f -L /Applications/RetroArch.app/Contents/Resources/cores/nestopia_libretro.dylib ~/Desktop/mario.nes`

and this shows the same effect: the command "hangs" in the shell. But I guess this runs as a kind of daemon which you connect to when you start retroarch over the OSX UI.

I found out that the switch -f starts the game without going into daemon mode first.

So this works: `/Applications/RetroArch.app/Contents/MacOS/RetroArch -f -L target/debug/libpinky_libretro.dylib ~/Desktop/mario.nes`. I was able to play the first level of Mario with sound, no lags, with the keyboard controller.

Maybe you could add the info we gathered in this issue to your readme?
